packages/nextcloud/lib/src/webdav/path_uri.dart

```diff
@@ -0,0 +1,154 @@
+    var isAbsolute = false;
+    var isDirectory = false;
+    final pathSegments = <String>[];
+
+    final parts = path.split('/');
+    if (parts.length == 1 && parts.single.isEmpty) {
+      return PathUri(
+        isAbsolute: false,
+        isDirectory: true,
+        pathSegments: BuiltList(),
+      );
+    }
+    for (var i = 0; i < parts.length; i++) {
+      final part = parts[i];
+      if (i == 0 && part.isEmpty) {
+        isAbsolute = true;
+      }
+      if (i == parts.length - 1 && part.isEmpty) {
+        isDirectory = true;
+      }
+      if (part.isNotEmpty) {
+        pathSegments.add(part);
+      }
+    }
+
+    return PathUri(
+      isAbsolute: isAbsolute,
+      isDirectory: isDirectory,
+      pathSegments: BuiltList(pathSegments),
+    );
+  }
+
+  /// Creates a new empty path URI representing the current working directory.
+  factory PathUri.cwd() => PathUri(
+        isAbsolute: false,
+        isDirectory: true,
+        pathSegments: BuiltList(),
+      );
+
+  /// Whether the path is an absolute path.
+  ///
+  /// If `true` [path] will start with a slash.
+  final bool isAbsolute;
+
+  /// Whether the path is a directory.
+  ///
+  /// If `true` [path] will end with a slash.
+  final bool isDirectory;
+
+  /// Returns the path as a list of its segments.
+  ///
+  /// See [path] for getting the path as a string.
+  final BuiltList<String> pathSegments;
+
+  /// Returns the path as a string.
+  ///
+  /// See [pathSegments] for getting the path as a list of its segments.
+  String get path {
+    if (pathSegments.isEmpty) {
+      if (isAbsolute) {
+        return '/';
+      }
+      return '';
+    }
+    final buffer = StringBuffer();
+    if (isAbsolute) {
+      buffer.write('/');
+    }
+    buffer.writeAll(pathSegments, '/');
+    if (isDirectory) {
+      buffer.write('/');
+    }
+    return buffer.toString();
+  }
+
+  /// Returns the name of the last element in path.
+  String get name => pathSegments.lastOrNull ?? '';
+
+  /// Returns the parent of the path.
+  PathUri? get parent {
+    if (pathSegments.isNotEmpty) {
+      return PathUri(
+        isAbsolute: isAbsolute,
+        isDirectory: true,
+        pathSegments: pathSegments.sublist(0, pathSegments.length - 1),
+      );
+    }
+
+    return null;
+  }
+
+  /// Joins the current path with another [path].
+  ///
+  /// If the current path is not a directory a [StateError] will be thrown.
+  /// See [isDirectory] for checking if the current path is a directory.
+  PathUri join(final PathUri other) {
+    if (!isDirectory) {
+      throw StateError('$this is not a directory.');
+    }
+
+    return PathUri(
+      isAbsolute: isAbsolute,
+      isDirectory: other.isDirectory,
+      pathSegments: BuiltList([
+        ...pathSegments,
+        ...other.pathSegments,
+      ]),
+    );
+  }
+
+  /// Renames the last path segment and returns a new path URI.
+  PathUri rename(final String name) => PathUri(
```

Comment:
I think that this is unsafe.
What if the new name is not sanitized and contains slashes?

We should probably throw in that case (needs better handling in the UI maybe with #998)

Reply:
This could also happen when creating a PathUri directly. Should we just remove all slashes and then remove all empty elements?